src/coordinator/functions/datetime/base.go

```diff
@@ -0,0 +1,133 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package datetime
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/m3db/m3db/src/coordinator/block"
+	"github.com/m3db/m3db/src/coordinator/executor/transform"
+	"github.com/m3db/m3db/src/coordinator/parser"
+)
+
+var emptyOp = BaseOp{}
+
+// BaseOp stores required properties for logical operations
+type BaseOp struct {
+	operatorType string
+	processorFn  makeProcessor
+}
+
+// OpType for the operator
+func (o BaseOp) OpType() string {
+	return o.operatorType
+}
+
+// String representation
+func (o BaseOp) String() string {
+	return fmt.Sprintf("type: %s", o.OpType())
+}
+
+// Node creates an execution node
+func (o BaseOp) Node(controller *transform.Controller) transform.OpNode {
+	return &baseNode{
+		controller: controller,
+		cache:      transform.NewBlockCache(),
+		op:         o,
+		processor:  o.processorFn(o, controller),
+	}
+}
+
+type baseNode struct {
+	op         BaseOp
+	controller *transform.Controller
+	cache      *transform.BlockCache
+	processor  Processor
+}
+
+// Ensure baseNode implements the types for lazy evaluation
+var _ transform.StepNode = (*baseNode)(nil)
+var _ transform.SeriesNode = (*baseNode)(nil)
+
+// ProcessStep allows step iteration
+func (c *baseNode) ProcessStep(step block.Step) (block.Step, error) {
+	processedValue := c.processor.ProcessStep(step.Values(), step.Time())
+	return block.NewColStep(step.Time(), processedValue), nil
+}
+
+// ProcessSeries allows series iteration
+func (c *baseNode) ProcessSeries(series block.Series, bounds block.Bounds) (block.Series, error) {
+	processedValue := c.processor.ProcessSeries(series.Values(), bounds)
+	return block.NewSeries(processedValue, series.Meta), nil
+}
+
+// Process the block
+func (c *baseNode) Process(ID parser.NodeID, b block.Block) error {
+	stepIter, err := b.StepIter()
+	if err != nil {
+		return err
+	}
+
+	builder, err := c.controller.BlockBuilder(stepIter.Meta(), stepIter.SeriesMeta())
+	if err != nil {
+		return err
+	}
+
+	if err := builder.AddCols(stepIter.StepCount()); err != nil {
+		return err
+	}
+
+	for index := 0; stepIter.Next(); index++ {
+		step, err := stepIter.Current()
+		if err != nil {
+			return err
+		}
+
+		values := c.processor.ProcessStep(step.Values(), step.Time())
+		for _, value := range values {
+			builder.AppendValue(index, value)
+		}
+	}
+
+	nextBlock := builder.Build()
+	defer nextBlock.Close()
+	return c.controller.Process(nextBlock)
+}
+
+// Meta returns the metadata for the block
+func (c *baseNode) Meta(meta block.Metadata) block.Metadata {
+	return meta
```

Comment:
is this correct ? Don't you need to modify some meta ?

Reply:
I don't think so, looks like the same meta data gets returned.